sim/shaman/enhancement/rotation.go

```diff
@@ -70,28 +68,41 @@ func (rotation *AdaptiveRotation) DoAction(enh *EnhancementShaman, sim *core.Sim
 		return
 	}
 
-	if enh.LavaburstWeave {
-		if enh.Talents.MaelstromWeapon > 0 && enh.MaelstromWeaponAura.GetStacks() >= 1 && enh.LavaBurst.IsReady(sim) {
-			lvbCastTime := enh.ApplyCastSpeed(enh.LavaBurst.DefaultCast.CastTime)
-			timeUntilSwing := enh.AutoAttacks.NextAttackAt() - sim.CurrentTime
-			if lvbCastTime < timeUntilSwing {
-				if !enh.LavaBurst.Cast(sim, target) {
-					enh.DoNothing()
-				}
-				return
-			}
-		}
-	}
-
-	if enh.Talents.MaelstromWeapon > 0 && enh.MaelstromWeaponAura.GetStacks() >= 1 {
-		lbCastTime := enh.LightningBolt.DefaultCast.CastTime - (time.Millisecond * time.Duration(500*enh.MaelstromWeaponAura.GetStacks()))
-		lbCastTime = enh.ApplyCastSpeed(lbCastTime)
+	if enh.MaelstromWeaponAura.GetStacks() >= 1 {
 		timeUntilSwing := enh.AutoAttacks.NextAttackAt() - sim.CurrentTime
+
 		if sim.CurrentTime > enh.AutoAttacks.NextAttackAt() {
 			timeUntilSwing = enh.AutoAttacks.MH.SwingDuration
 		}
-		if lbCastTime < timeUntilSwing {
-			if !enh.LightningBolt.Cast(sim, target) {
+
+		swingSpeed := core.MinDuration(enh.AutoAttacks.MainhandSwingSpeed(), enh.AutoAttacks.OffhandSwingSpeed())
+		swingSpeed = core.MaxDuration(swingSpeed, timeUntilSwing)
+		latency := time.Duration(enh.WeaveLatency) * time.Millisecond
+		swingSpeed -= latency
+		latency = core.MaxDuration(0, timeUntilSwing-swingSpeed)
+
+		var spellToCast *core.Spell
+		castTime := time.Duration(0)
+
+		if enh.LavaburstWeave && enh.LavaBurst.IsReady(sim) {
+			castTime = enh.ApplyCastSpeed(enh.LavaBurst.DefaultCast.CastTime) + latency
+			spellToCast = enh.LavaBurst
+		}
+
+		// If we can't fit a Lava Burst in try a Lightning Bolt
+		if castTime >= timeUntilSwing || spellToCast == nil {
+			castTime = enh.LightningBolt.DefaultCast.CastTime - (time.Millisecond * time.Duration(500*enh.MaelstromWeaponAura.GetStacks()))
+			castTime = enh.ApplyCastSpeed(castTime) + latency
+			spellToCast = enh.LightningBolt
+		}
+
+		if castTime < timeUntilSwing {
+			if latency > 0 {
+				enh.HardcastWaitUntil(sim, sim.CurrentTime+latency, func(sim *core.Simulation, _ *core.Unit) {
+					spellToCast.Cast(sim, target)
+				})
+				enh.DoNothing()
+			} else if !spellToCast.Cast(sim, target) {
 				enh.WaitForMana(sim, enh.LightningBolt.CurCast.Cost)
 			}
 			return
```

Comment:
all of this is gonna need big changes to work with some of the other stuff thats been done and is being worked on. probably gonna go through and rework a lot of rotation code today, unfortunately its a bit tricky because like 4 people are trying to make changes to the rotation at once

Reply:
yep imagine this will get refactored also, the *spellToCast will prob just go away. the calculate the latency and prob the HardcastWaitUntil() might still get used which is really the meat the other stuff is just organization.